json-decode.go

```diff
@@ -210,46 +202,12 @@ func (cdc *Codec) decodeReflectJSONArrayOrSlice(bz []byte, info *TypeInfo, rv re
 	return nil
 }
 
-func safeNil(rv reflect.Value) bool {
-	switch rv.Kind() {
-	case reflect.Chan, reflect.Func, reflect.Interface, reflect.Map, reflect.Ptr, reflect.Slice:
-		return rv.IsNil()
-	default:
-		return false
-	}
-}
-
-// processIfUnmarshaler checks if the type or its pointer
-// to implements json.Unmarshaler and if so, invokes it.
-func processIfUnmarshaler(bz []byte, rv reflect.Value) (unmarshalable bool, err error) {
-	if rv.Type().Implements(unmarshalerType) {
-		if safeNil(rv) {
-			return false, nil
-		}
-		return true, rv.Interface().(json.Unmarshaler).UnmarshalJSON(bz)
-	} else if itsPtrImplements(rv, unmarshalerType) {
-		rvPtr := rv.Addr()
-		// Otherwise check if its pointer implements it
-		if err := rvPtr.Interface().(json.Unmarshaler).UnmarshalJSON(bz); err != nil {
-			return true, err
-		}
-		rv.Set(rvPtr.Elem())
-		return true, nil
-	}
-
-	return false, nil
-}
-
 func (cdc *Codec) decodeReflectJSONStruct(bz []byte, info *TypeInfo, rv reflect.Value, opts FieldOptions) error {
 	nf := len(info.Fields)
 	if nf == 0 {
 		return nil
 	}
 
-	if ok, err := processIfUnmarshaler(bz, rv); ok || err != nil {
```

Comment:
I don't think this is needed here... it should be checked once at the decodeReflectJSON func.